### PR TITLE
Restructure File.read into the new design

### PR DESCRIPTION
**Pull request by jdunkerley**
[ci no changelog needed]

### Pull Request Description

Changes following Marcin's work. Should be back to very similar public API as before.

- Add an "interface" type: `Standard.Base.System.File_Format.File_Format`.
- All `File_Format` types now have a `can_read` method to decide if they can read a file.
- Move `Standard.Table.IO.File_Format.Text.Text_Data` to `Standard.Base.System.File_Format.Plain_Text_Format.Plain_Text`.
- Move `Standard.Table.IO.File_Format.Bytes` to `Standard.Base.System.File_Format.Bytes`.
- Move `Standard.Table.IO.File_Format.Infer` to `Standard.Base.System.File_Format.Infer`. **(doesn't belong here...)**
- Move `Standard.Table.IO.File_Format.Unsupported_File_Type` to `Standard.Base.Error.Common.Unsupported_File_Type`.
- Add `Infer`, `File_Format`, `Bytes`, `Plain_Text`, `Plain_Text_Format` to `Standard.Base` exports.
- Fold extension methods of `Standard.Base.Meta.Unresolved_Symbol` into type.
- Move `Standard.Table.IO.File_Format.Auto` to `Standard.Table.IO.Auto_Detect.Auto_Detect`.
  - Added a `types` Vector of all the built in formats.
  - `Auto_Detect` asks each type if they `can_read` a file.
- Broke up and moved `Standard.Table.IO.Excel` into `Standard.Table.Excel`:
  - Moved `Standard.Table.IO.File_Format.Excel.Excel_Data` to `Standard.Table.Excel.Excel_Format.Excel_Format.Excel`. 
  - Renamed `Sheet` to `Worksheet`.
  - Internal types `Reader` and `Writer` providing the actual read and write methods.
- Created `Standard.Table.Delimited` with similar structure to `Standard.Table.Excel`:
  - Moved `Standard.Table.IO.File_Format.Delimited.Delimited_Data` to `Standard.Table.Delimited.Delimited_Format.Delimited_Format.Delimited`.
  - Moved `Standard.Table.IO.Quote_Style` to `Standard.Table.Delimited.Quote_Style`.
  - Moved the `Reader` and `Writer` internal types into here. Renamed methods to have unique names.
- Add `Aggregate_Column`, `Auto_Detect`, `Delimited`, `Delimited_Format`, `Excel`, `Excel_Format`, `Sheet_Names`, `Range_Names`, `Worksheet` and `Cell_Range` to `Standard.Table` exports.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from Standard.Base import all
+from Standard.Base.Error.Common import unimplemented
+
+type File_Format
+    ## If the File_Format supports reading from the file, return a configured instance.
+    can_read : File -> Any
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso]:
```suggestion
    can_read : File -> Any | Nothing
```

I know that semantically `Any | Nothing =:= Any`, but I suggest it may be worth including it explicitly here to clearly mark that `Nothing` means "cannot read".

Still, I'd update the docstring to say what is returned if it doesn't support reading - because currently it is not explicitly stated.

I'm thinking that the name `can_read` is a bit confusing if it not only checks that but also returns an instance. But honestly I'm not sure if I have a better idea, so probably let's juts keep it. Maybe `configure_format_for`? Doesn't sound too great though.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 2]
```diff
@@ -960,13 +956,13 @@ type Table
...
 
-         - `File_Format.Auto`: The file format is determined by the file
-           extension of the path argument.
-         - `File_Format.Bytes` and `File_Format.Text`: The Table does not
-           support these types in the `write` function. If passed as format, an
-           `Illegal_Argument_Error` is raised. To write out the table as plain
-           text, the user needs to call the `Text.from Table` method and then
-           use the `Text.write` function.
+         - `Auto_Detect`: The file format is determined by the file extension
+            of the path argument.
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso]: I would not limit ourselves to determining by extension only. Currently that is the case, but I can imagine situations where the format may need to read a few bytes from the file to determine if it can support it.
Do you think such scenarios are feasible in the future? If yes, I'd rephrase this to say just something that the file format is determined automatically based on the provided file?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Delimited/Reader.enso [hunk 3]
```diff
@@ -35,17 +35,17 @@ polyglot java import org.enso.table.read.QuoteStrippingParser
      operation. By default, a warning is issued, but the operation proceeds.
      If set to `Report_Error`, the operation fails with a dataflow error.
      If set to `Ignore`, the operation proceeds without errors or warnings.
-read_file : Delimited -> File -> Problem_Behavior -> Any
-read_file format file on_problems =
+read_delimited_file : Delimited_Format -> File -> Problem_Behavior -> Any
+read_delimited_file format file on_problems =
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Delimited/Reader.enso]: I'm a bit unsure about these renames, it is a stylistic convention so I'm not gonna argue too much, it is ok too.

But what about not importing the functions directly into scope but using them as qualified?
For example I think we could even do:
```
import Standard.Table.Delimited

...
Delimited.Reader.read_file ...
```

I really like the look of that and there is actually _less_ scope pollution.

Although the current one is ok too, I think I just wanted to put another one on the table so we can choose one.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Format.enso [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import Standard.Table.Data.Table
+
+from Standard.Table.Excel.Section import Excel_Section, Worksheet, Sheet_Names, Range_Names
+from Standard.Table.Excel.Reader import read_excel
+from Standard.Table.Excel.Writer import write_excel
+
+## PRIVATE
+   Resolve the xls_format setting to a boolean.
+is_xls_format : (Boolean|Infer) -> File -> Boolean
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Format.enso]: Nitpick but `is_xls_format True (File.new "my_TEXT_file.txt") == True` - looks a bit confusing - wait 'txt' is not XLS.
Maybe we can call it `should_treat_as_xls_format`? Then it will express the intent a bit clearer IMO.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Reader.enso [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        Worksheet sheet skip_rows row_limit ->
+            prepare_reader_table on_problems <| case sheet of
+                Integer -> ExcelReader.readSheetByIndex stream sheet (make_java_headers headers) skip_rows row_limit xls_format
+                Text -> ExcelReader.readSheetByName stream sheet (make_java_headers headers) skip_rows row_limit xls_format
+        Cell_Range address skip_rows row_limit ->
+            prepare_reader_table on_problems <| case address of
+                Excel_Range_Data _ -> ExcelReader.readRange stream address.java_range (make_java_headers headers) skip_rows row_limit xls_format
+                Text -> ExcelReader.readRangeByName stream address (make_java_headers headers) skip_rows row_limit xls_format
+
+    handle_reader file reader
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Reader.enso]: Please add a newline at EOF

File: distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso [hunk 6]
```diff
@@ -0,0 +1,36 @@
+from Standard.Base import Any, Problem_Behavior, Nothing, Error, Panic, Meta, File, File_Format, Plain_Text, Bytes
+from Standard.Base.Error.Common import Unsupported_File_Type_Data, No_Such_Method_Error_Data, Illegal_Argument_Error_Data
+
+from Standard.Table.Delimited.Delimited_Format import Delimited
+from Standard.Table.Excel.Excel_Format import Excel
+
+types = [Plain_Text, Delimited ',', Excel, Bytes]
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso]:
```suggestion
## PRIVATE
known_file_types = [Plain_Text, Delimited ',', Excel, Bytes]
```

I'd make the name a bit more meaningful, and I think we should either add a comment explaining what this is OR a comment `PRIVATE` so it is kept private.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Delimited/Delimited_Format.enso [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+         defaults to `Nothing` which means that comments are disabled.
+    Delimited (delimiter:Text) (encoding:Encoding=Encoding.utf_8) (skip_rows:Integer=0) (row_limit:Integer|Nothing=Nothing) (quote_style:Quote_Style=Quote_Style.With_Quotes) (headers:Boolean|Infer=Infer) (value_formatter:Data_Formatter|Nothing=Data_Formatter_Data) (keep_invalid_rows:Boolean=True) (line_endings:Line_Ending_Style=Infer) (comment_character:Text|Nothing=Nothing)
+
+    ## If the File_Format supports reading from the file, return a configured instance.
+    can_read : File -> Any
+    can_read self file =
+        case file.extension of
+            ".csv" -> Delimited ','
+            ".tsv" -> Delimited '\t'
+            _ -> Nothing
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Delimited/Delimited_Format.enso]: I'm not sure about this.

Shouldn't we instead do it so that `Delimited ','` supports `csv` and `Delimited '\t'` supports `tsv` - and just put two instances of these in our supported types list?
It seems a bit weird to me that we construct `Delimited ','` but then it answers positively to `can_read "foo.tsv"` but will yield a changed configuration. Shouldn't we instead just have these as two kind-of-separate file types? And maybe then we could get rid of this pre-configuration logic and just make `can_read` return a `Boolean`?

Unless you envision this more complex logic should prove more useful in the future - which I think is quite possible indeed. Still I'm a bit uneasy about the fact that we need to instantiate an essentially fully configured file type (like `Delimited`) instance to then infer a possibly another instance. Feels to me like mixing responsibilities a bit - it seems that we may need to have two separate concepts - one for detecting the default configuration for a given set of formats and another for holding a concrete configuration after it was detected or configured manually. So maybe this `can_read` should be defined as a static on `Delimited_Format` instead?

Even if we want to keep the above as-is, at the very least I think we should do 
```suggestion
    can_read : File -> Any
    can_read self file =
        case file.extension of
            ".csv" -> self.clone delimiter=','
            ".tsv" -> self.clone delimiter='\t'
            _ -> Nothing
```

(I know that `clone` does not currently support overridding the `delimiter` but I think we can easily adapt it).

Right now we can pass a `Delimited` instance with some more complex configuration, ask it "can you read X" and we get back a new instance - with all of our configuration wiped out. So shall we still use the original one? But the delimiter might have changed, so we probably shouldn't. But the new one does not have all of our overrides.
Of course for now we don't have such uses, because we don't use more complex configs together with `can_read` - well that just suggests me that `can_read` is sitting in wrong place and the configuration and detection are separate responsibilities that should not be coupled in this single concept.

**Review by radeusgd — COMMENTED** [review]
Still in progress of reviewing, but I already have a few major comments on handling file type detection. Can we discuss the approach a bit?

File: distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+type Auto_Detect
+    ## If the File_Format supports reading from the file, return a configured instance.
+    can_read : File -> Any
+    can_read self file =
+        reader idx =
+            formatter = types.at idx . can_read file
+            if formatter.is_nothing.not then formatter else
+                if idx + 1 == types.length then Nothing else
+                    @Tail_Call reader (idx + 1)
+        reader 0
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso]: This feels a bit like we should have a relevant function in `Vector` instead of writing the loop here, although I'm not sure what this function should be called. We can always do `types.map .can_read . find (x -> x.is_nothing.not)` - but as our `map` is not lazy this will not terminate early.

If we want to keep this function, can we do
```suggestion
        reader idx =
            if idx >= types.length then Nothing else
                formatter = types.at idx . can_read file
                if formatter.is_nothing.not then formatter else
                    @Tail_Call reader (idx + 1)
        reader 0
```

I prefer the length check at the beginning in case the `types` were empty (I know they aren't but still...) and then the aritmetic is simpler (no idx+1) and its more conistent with other loops like that that we write - which I think is nice because you get used to such patterns and in time I recognize such patterns almost automatically. Whereas with the other ordering I looked at the loop and had to think a bit more to see what is going on.

File: distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    read self file on_problems =
+        reader = self.can_read file
+        if reader == Nothing then Error.throw (Unsupported_File_Type_Data ("No File_Format supports '" + file.extension + "'")) else
+            reader.read file on_problems
+
+    ## PRIVATE
+       Implements the `Table.write` for this `File_Format`.
+    write_table : File -> Table -> Existing_File_Behavior -> Match_Columns -> Problem_Behavior -> Nothing
+    write_table self file table on_existing_file match_columns on_problems =
+        format = self.can_read file
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso]: If in the future `can_read` would want to not only rely on extension but also contents, this logic will not work anymore as in `write` case the file may not yet exist. But for now I guess it's ok, we can always revisit when need arises.

**Review by radeusgd — COMMENTED** [review]
The overall shape looks all good to me, but I think I'd like to have another look after we address the discussion on file format detection being coupled with config.

File: distribution/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso [hunk 10]
```diff
@@ -15,8 +11,8 @@ type Plain_Text_Format
     Plain_Text (encoding:Encoding=Encoding.utf_8)
 
     ## If the File_Format supports reading from the file, return a configured instance.
-    can_read : File -> Any
-    can_read self file =
+    for_file : File -> Any | Nothing
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso]: Nitpick, but we can actually make the signature more precise:
```suggestion
    for_file : File -> File_Format | Nothing
```

and then in `Delimited_Format` we can have `for_file : File -> Delimited_Format | Nothing`. I don't think we need the return types of both of these variants to be exactly the same so I think its worth refining them.

File: distribution/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso [hunk 11]
```diff
@@ -30,8 +26,8 @@ type Plain_Text_Format
 
 type Bytes
     ## If the File_Format supports reading from the file, return a configured instance.
-    can_read : File -> Any
-    can_read self file =
+    for_file : File -> Any | Nothing
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso]:
```suggestion
    for_file : File -> Bytes | Nothing
```

File: distribution/lib/Standard/Table/0.0.0-dev/src/Delimited/Delimited_Format.enso [hunk 12]
```diff
@@ -2,13 +2,21 @@ from Standard.Base import all
...
 
 import Standard.Table.Delimited.Quote_Style
 
+## If the File_Format supports reading from the file, return a configured instance.
+for_file : File -> Any | Nothing
+for_file file =
+    case file.extension of
+        ".csv" -> Delimited ','
+        ".tsv" -> Delimited '\t'
+        _ -> Nothing
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Delimited/Delimited_Format.enso]: Cannot we define it as a static on the `Delimited_Format` type instead of module? Or doesn't it work?

If it doesn't work, maybe we can add a TODO to move it to the type at some point?
Or is it an imports issue, i.e. we'd then need to do `Delimited_Format.Delimited_Format` or sth? 

(This is actually a deeper issue that we need to keep in mind - once we are using types more - a lot of our types are sitting in files (modules) with the same name and we don't necessarily want to be calling `XYZ.XYZ` everywhere)

File: distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso [hunk 13]
```diff
@@ -1,36 +1,39 @@
...
 
-from Standard.Table.Delimited.Delimited_Format import Delimited
-from Standard.Table.Excel.Excel_Format import Excel
+import Standard.Table.Delimited.Delimited_Format
+import Standard.Table.Excel.Excel_Format
 
-types = [Plain_Text, Delimited ',', Excel, Bytes]
+## PRIVATE
+   Set of methods to get the configured File_Format for a file.
+format_providers = [Plain_Text_Format.for_file, Bytes.for_file, Delimited_Format.for_file, Excel_Format.for_file]
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso]: This is a very minor thing but I'm thinking if we want to have `XYZ.for_file` here or just `XYZ` and make sure that `get_format` just delegates to `for_file` assuming a consistent interface?

An argument for _not_ having functions but modules instead on this list is that in the future we may introduce another function to the provider "interface" (oh how good it would be to have actual interfaces...) and then having modules here will naturally allow us to use this other method. With keeping callbacks, we don't have this kind of space for extensions. It also seems more natural to me with the plans for registering the providers - I'd rather register a provider instance than a callback function.

**Review by radeusgd — APPROVED** [review]
Looks good, thanks for the changes.

A few more "low-priority" suggestions.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Section.enso [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+    ## Gets the data from a specific sheet. Column names are the Excel column
+       names.
+    Worksheet (sheet:(Integer|Text)=1) (skip_rows:Integer=0) (row_limit:(Integer|Nothing)=Nothing)
+
+    ## Gets a specific range (taking either a defined name or external style
+       address) from the workbook.
+       If it is a single cell, it will be treated as the top left cell and will
+       expand right and down to cover the connected cells.
+    Cell_Range (address:(Text|Excel_Range)) (skip_rows:Integer=0) (row_limit:(Integer|Nothing)=Nothing)
```
Comment (hubertp) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Section.enso]:
```suggestion
    Cell_Range (address:(Text|Excel_Range)) (skip_rows:Integer=0) (row_limit:(Integer|Nothing)=Nothing)

```

**Review by hubertp — APPROVED** [review]
lgtm
